### PR TITLE
Add FastFinality contract #106

### LINKED ISSUE #106: Research Fast Finality
- The operator need 1x+const of deposited bond
- A transaction has time range parameter

If a user has an unsubmitted signed transaction, the user can send the include request transaction on the root chain.
There are 2 patterns.

- the user is honest
- the user is not honest

If the operator includes the transaction into a block after that, this request was will be challenged.
If no, the operator will be penalized.

### PR DESCRIPTION
**Pull request by syuhei176**
## Task List

- [x] dispute
- [x] challenge
- [x] secondDispute
- [x] finalizeDispute
- [x] spec: dispute
- [x] spec: challenge
- [x] spec: secondDispute
- [x] spec: finalizeDispute

Closes #106

### CODE REVIEW

**Comment by syuhei176** [comment]
I found a problem.

Basically, sender and operator are malicious in Plasma double spend attack scenario.
And FF transactions were signed by sender and operator.

Sender sign two different transactions, tx1 is changing owner to the receiver, tx2 is to the operator.

sender send tx1 to the receiver, but the operator includes tx2 in the block and include tx1 in the next block.  Then tx1 is double spent transaction. But receiver already receives tx1, and the sender gets an item from the receiver.

The receiver sends a dispute request by tx1, but the operator can challenge by inclusion proof of tx1.

To solve this.
**The operator should exit tx1 first for the receiver.**

File: packages/chamber-rootchain/contracts/FastFinality.sol [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    uint256 index,
+    uint256 blkNum,
+    bytes txBytes,
+    bytes txInfos
+  )
+    public
+    returns (bool)
+  {
+    // check inclusion
+    rootchain.call(
```
Comment (syuhei176) [inline packages/chamber-rootchain/contracts/FastFinality.sol]: If operator and receiver swap their coins, the receiver will have exitable transaction.

**Comment by syuhei176** [comment]
> To solve this.
> The operator should show that tx1 is exitable for the receiver before challenging the dispute.

The operator can show that tx1 is exitable by atomic swap.

**Comment by syuhei176** [comment]
> > To solve this.
> > The operator should show that tx1 is exitable for the receiver before challenging the dispute.
> 
> The operator can show that tx1 is exitable by atomic swap.

The operator needs receiver's signature, that's not good.

**Comment by syuhei176** [comment]
I wrote full flow chart for FF.

File: packages/chamber-rootchain/contracts/FastFinality.sol [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  /**
+    * @dev 
+    */
+  function challenge(
+    uint disputePos,
+    uint256 index,
+    uint256 blkNum,
+    bytes txBytes,
+    bytes txInfos
```
Comment (shogochiai) [inline packages/chamber-rootchain/contracts/FastFinality.sol]: `_txInfos` data structure is being used in many placed but structure information is somehow inaccessible. Please note here how should it be.

Comment (syuhei176) [reply packages/chamber-rootchain/contracts/FastFinality.sol]: txInfos include proof, signature and confsig.

Comment (shogochiai) [reply packages/chamber-rootchain/contracts/FastFinality.sol]: It's okay to use RLP bytes I also believe :)
But format indication is better to be this time. I needed 2~3 hops to reach that.

Comment (syuhei176) [reply packages/chamber-rootchain/contracts/FastFinality.sol]: Added comments as js doc;)

**Comment by syuhei176** [comment]
I added the second dispute.

**Comment by syuhei176** [comment]
added all specs.

File: packages/chamber-rootchain/contracts/FastFinality.sol [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+contract FastFinality {
+
+  /*
+    * Storage
+    */
+  struct Dispute {
+    address recipient;
+    uint256 withdrawableAt;
+    uint256 amount;
+    bool isUsed;
```
[the remark this answers is not in the document]
Comment (syuhei176) [reply packages/chamber-rootchain/contracts/FastFinality.sol]: Yes, I'll add bond.

[the remark this answers is not in the document]
Comment (syuhei176) [reply packages/chamber-rootchain/contracts/FastFinality.sol]: added https://github.com/cryptoeconomicslab/plasma-chamber/pull/160/files#diff-7ba7fa82cf23f12517bb38340be2a897R24

**Comment by syuhei176** [comment]
It will be merged into https://github.com/cryptoeconomicslab/chamber-packages/tree/master/packages/chamber-contracts

**Pull request merged**